File: aoc/2023/aoc202306.hs
import Data.Char(isDigit)

parse :: String -> [(Int,Int)]
parse = join . map (map read . drop 1 . words) . lines
  where join [time,distance] = zip time distance

-- Let t be the race time.  Let h be the hold time.
-- Let s be the speed.  s = h.
-- Let d be the distance.  d = s⋅(t-h) = th - h^2.
-- Given a distance, there are two hold times
--   h = (t ± √(t^2 - 4d))/2

ways :: (Int,Int) -> Int
ways (t,d)
  | (floor (sqrt (fromIntegral disc)))^2 == disc && even (t + floor (sqrt (fromIntegral disc))) = high - low - 1
  | otherwise = high - low + 1
  where
    disc = t^2 - 4*d
    high = floor ((fromIntegral t + sqrt (fromIntegral disc))/2)
    low = ceiling ((fromIntegral t - sqrt (fromIntegral disc))/2)

result :: String -> Int
result = product . map ways . parse

testData :: String
testData = unlines [
    "Time:      7  15   30",
    "Distance:  9  40  200"
    ]

test :: ()
test
  | map ways (parse testData) /= [4,8,9] = error "a"
  | result testData /= 288 = error "a"
  | otherwise = ()
  where

part1 :: IO Int
part1 = fmap result $ readFile "input/06.txt"

parse2 :: String -> (Int,Int)
parse2 = join . map (read . filter isDigit) . lines
  where join [time,distance] = (time,distance)

result2 :: String -> Int
result2 = ways . parse2

test2 :: ()
test2
  | result2 testData /= 71503 = error "a"
  | otherwise = ()

part2 :: IO Int
part2 = fmap result2 $ readFile "input/06.txt"

File: aoc/2025/aoc202306.go
package main

import (
	"math"
)

func init() {
	Register(&aoc202306{
		AOC: AOC{
			Day:           6,
			InputFilename: "../2023/input/06.txt",
			Tests: []Test{
				Test{
					Input: `Time:      7  15   30
Distance:  9  40  200
`,
					Part1: "288",
					Part2: "71503",
				},
			},
		},
	})
}

type aoc202306 struct {
	AOC
}

func (aoc *aoc202306) parse(input *Input) [][2]int {
	numbers := input.Ints()
	races := [][2]int{}
	for i := range len(numbers) / 2 {
		races = append(races, [2]int{numbers[i], numbers[i+len(numbers)/2]})
	}
	return races
}

func (aoc *aoc202306) ways(t, d int) int {
	// d = (t-p)*p
	// p = t/2 +- sqrt(t^2 - 4*d)/2
	disc := math.Sqrt(float64(t*t - 4*d))
	plo := int(math.Ceil((float64(t) - disc) / 2))
	if (t-plo)*plo <= d {
		plo++
	}
	phi := int(math.Floor((float64(t) + disc) / 2))
	if (t-phi)*phi <= d {
		phi--
	}
	return phi - plo + 1
}

func (aoc *aoc202306) Part1(input *Input) string {
	result := 1
	for _, race := range aoc.parse(input) {
		result *= aoc.ways(race[0], race[1])
	}
	return IntResult(result)
}

func (aoc *aoc202306) parse2(input *Input) (int, int) {
	t := 0
	d := 0
	haveT := false
	for ch, ok := input.Char(); ok; ch, ok = input.Char() {
		switch ch {
		case '\n':
			haveT = true
		case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
			if !haveT {
				t = t*10 + int(ch-'0')
			} else {
				d = d*10 + int(ch-'0')
			}
		}
	}
	return t, d
}

func (aoc *aoc202306) Part2(input *Input) string {
	return IntResult(aoc.ways(aoc.parse2(input)))
}
